src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+                    symbolStartContext.RegisterOperationAction(AnalyzeInvocationOperation, OperationKind.Invocation);
+                    symbolStartContext.RegisterSymbolEndAction(OnSymbolEnd);
+                }, SymbolKind.NamedType);
+
+                return;
+
+                // Local functions.
+                void AnalyzeSymbolDeclaration(SymbolAnalysisContext symbolContext)
+                {
+                    if (IsCandidateSymbol(symbolContext.Symbol))
+                    {
+                        // Initialize unused state to 'true'.
+                        symbolValueUsageStateMap.GetOrAdd(symbolContext.Symbol, valueFactory: s => ValueUsageInfo.None);
+                    }
+                }
+
+                void OnSymbolUsage(ISymbol memberSymbol, ValueUsageInfo usageInfo)
+                {
+                    Debug.Assert(IsCandidateSymbol(memberSymbol));
+
+                    // Update the usage info for the memberSymbol
+                    symbolValueUsageStateMap.AddOrUpdate(
+                        memberSymbol,
+                        addValue: usageInfo,
+                        updateValueFactory: (sym, currentUsageInfo) => currentUsageInfo | usageInfo);
+                }
+
+                void AnalyzeMemberReferenceOperation(OperationAnalysisContext operationContext)
+                {
+                    var memberReference = (IMemberReferenceOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(memberReference.Member))
+                    {
+                        // Get the value usage info.
+                        var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                        // Special case for VB
+                        // Compound assigment is a statement in VB that does not return a value.
+                        // So, we treat this usage as a write-only usage.
+                        if (memberReference.Language == LanguageNames.VisualBasic &&
+                            memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference)
+                        {
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+
+                        OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                    }
+                }
+
+                void AnalyzeInvocationOperation(OperationAnalysisContext operationContext)
+                {
+                    var invocation = (IInvocationOperation)operationContext.Operation;
+                    if (IsCandidateSymbol(invocation.TargetMethod))
+                    {
+                        OnSymbolUsage(invocation.TargetMethod, ValueUsageInfo.Read);
+                    }
```

Comment:
probably worth explaining why we need this hwen we have AnalyzeMemberReferenceOperation

Reply:
Added a note about requiring to do this due to [https://github.com/dotnet/roslyn/issues/26206](https://github.com/dotnet/roslyn/issues/26206)

---
In reply to: [212731300](https://github.com/dotnet/roslyn/pull/29511#discussion_r212731300) [](ancestors = 212731300)